File: examples/35-virtualization-and-kvm/go/main.go
// overheadbench: three host-local microbenchmarks used to talk about
// virtualization/container overhead in chapter 35 —
//
//	(a) syscall latency: a tight getpid(2) loop, ns/call
//	(b) memory bandwidth: sequential read of a big buffer, GB/s
//	(c) small-file IO: create/write/fsync/unlink loop, ops/s
//
//	overheadbench [--bench syscall|mem|io|all] [--iters N]
//
// Prints one line per bench: "bench=<name> metric=<value> unit=<unit>".
// This binary always measures wherever it runs (host, VM, or container) —
// the chapter runs the *same* binary in all three places and tabulates the
// numbers; this example itself only asserts the host numbers are sane.
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"golang.org/x/sys/unix"
)

const usage = "usage: overheadbench [--bench syscall|mem|io|all] [--iters N]"

// Defaults are chosen so each bench runs for roughly tens to a few hundred
// milliseconds on a modern host; identical across all three languages.
const (
	syscallDefaultIters = uint64(200_000)
	memDefaultPasses    = uint64(16)
	memBufBytes         = uint64(128) * 1024 * 1024 // 128 MiB
	ioDefaultIters      = uint64(200)
	ioFileBytes         = 4096
)

type bench int

const (
	benchSyscall bench = iota
	benchMem
	benchIo
	benchAll
)

type config struct {
	bench bench
	iters *uint64
}

func parseArgs(args []string) (config, error) {
	cfg := config{bench: benchAll}
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--bench":
			i++
			if i >= len(args) {
				return config{}, fmt.Errorf("--bench requires a value")
			}
			switch args[i] {
			case "syscall":
				cfg.bench = benchSyscall
			case "mem":
				cfg.bench = benchMem
			case "io":
				cfg.bench = benchIo
			case "all":
				cfg.bench = benchAll
			default:
				return config{}, fmt.Errorf("unknown --bench value: %s", args[i])
			}
		case "--iters":
			i++
			if i >= len(args) {
				return config{}, fmt.Errorf("--iters requires a value")
			}
			n, err := strconv.ParseUint(args[i], 10, 64)
			if err != nil || n == 0 {
				return config{}, fmt.Errorf("--iters value must be a positive integer: %s", args[i])
			}
			cfg.iters = &n
		default:
			return config{}, fmt.Errorf("unknown argument: %s", args[i])
		}
	}
	return cfg, nil
}

// benchSyscallRun: tight getpid(2) loop, ns/call.
func benchSyscallRun(iters uint64) float64 {
	start := time.Now()
	for i := uint64(0); i < iters; i++ {
		// unix.Getpid issues the raw getpid(2) syscall on every call (the Go
		// runtime does not cache it) — every iteration really crosses into
		// the kernel, the syscall-boundary cost this chapter contrasts
		// against VM (vmexit trap) and container (near-zero extra) overhead.
		unix.Getpid()
	}
	ns := float64(time.Since(start).Nanoseconds())
	if ns <= 0 {
		ns = 1
	}
	return ns / float64(iters)
}

// benchMemRun: sequential read bandwidth over a big buffer, GB/s.
func benchMemRun(passes uint64) float64 {
	words := memBufBytes / 8
	buf := make([]uint64, words)
	for i := range buf {
		buf[i] = uint64(i) * 2654435761
	}

	var sum uint64
	start := time.Now()
	for p := uint64(0); p < passes; p++ {
		for _, w := range buf {
			sum += w
		}
	}
	secs := time.Since(start).Seconds()
	if secs <= 0 {
		secs = 1e-9
	}

	// Consume `sum` after the clock stops so the compiler cannot prove the
	// whole scan is dead and discard it; this doesn't affect the measured time.
	if sum == 0xFFFFFFFFFFFFFFFF {
		fmt.Fprintln(os.Stderr, "unreachable sink:", sum)
	}

	bytes := float64(memBufBytes) * float64(passes)
	return bytes / secs / 1e9
}

// benchIoRun: create/write/fsync/unlink loop, ops/s.
func benchIoRun(iters uint64) (float64, error) {
	// Relative to the CWD (the demo.sh working directory), not the system
	// temp dir: on many dev hosts that's tmpfs, where fsync is nearly free
	// and the number stops meaning anything as "disk IO overhead". The
	// example directory itself is normally on a real filesystem, which is
	// the point of comparing this number across host/VM/container.
	dir, err := os.MkdirTemp(".", "overheadbench-io-*")
	if err != nil {
		return 0, fmt.Errorf("mkdirtemp: %w", err)
	}
	defer os.RemoveAll(dir)

	path := dir + "/probe"
	payload := make([]byte, ioFileBytes)
	for i := range payload {
		payload[i] = 'x'
	}

	start := time.Now()
	for i := uint64(0); i < iters; i++ {
		f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o600)
		if err != nil {
			return 0, fmt.Errorf("open: %w", err)
		}
		n, err := f.Write(payload)
		if err != nil || n != len(payload) {
			f.Close()
			return 0, fmt.Errorf("write: %w", err)
		}
		if err := f.Sync(); err != nil {
			f.Close()
			return 0, fmt.Errorf("fsync: %w", err)
		}
		if err := f.Close(); err != nil {
			return 0, fmt.Errorf("close: %w", err)
		}
		if err := os.Remove(path); err != nil {
			return 0, fmt.Errorf("unlink: %w", err)
		}
	}
	secs := time.Since(start).Seconds()
	if secs <= 0 {
		secs = 1e-9
	}
	return float64(iters) / secs, nil
}

type row struct {
	name  string
	unit  string
	value float64
}

func runBenches(cfg config) ([]row, error) {
	wantSyscall := cfg.bench == benchSyscall || cfg.bench == benchAll
	wantMem := cfg.bench == benchMem || cfg.bench == benchAll
	wantIo := cfg.bench == benchIo || cfg.bench == benchAll

	var rows []row
	if wantSyscall {
		iters := syscallDefaultIters
		if cfg.iters != nil {
			iters = *cfg.iters
		}
		rows = append(rows, row{"syscall", "ns/call", benchSyscallRun(iters)})
	}
	if wantMem {
		passes := memDefaultPasses
		if cfg.iters != nil {
			passes = *cfg.iters
		}
		rows = append(rows, row{"mem", "GB/s", benchMemRun(passes)})
	}
	if wantIo {
		iters := ioDefaultIters
		if cfg.iters != nil {
			iters = *cfg.iters
		}
		v, err := benchIoRun(iters)
		if err != nil {
			return nil, fmt.Errorf("io bench failed: %w", err)
		}
		rows = append(rows, row{"io", "ops/s", v})
	}
	return rows, nil
}

func main() {
	cfg, err := parseArgs(os.Args[1:])
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		fmt.Fprintln(os.Stderr, usage)
		os.Exit(2)
	}

	rows, err := runBenches(cfg)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	for _, r := range rows {
		fmt.Printf("bench=%s metric=%.2f unit=%s\n", r.name, r.value, r.unit)
	}
}

File: examples/35-virtualization-and-kvm/rust/src/main.rs
// overheadbench: three host-local microbenchmarks used to talk about
// virtualization/container overhead in chapter 35 —
//   (a) syscall latency: a tight getpid(2) loop, ns/call
//   (b) memory bandwidth: sequential read of a big buffer, GB/s
//   (c) small-file IO: create/write/fsync/unlink loop, ops/s
//
//   overheadbench [--bench syscall|mem|io|all] [--iters N]
//
// Prints one line per bench: "bench=<name> metric=<value> unit=<unit>".
// This binary always measures wherever it runs (host, VM, or container) —
// the chapter runs the *same* binary in all three places and tabulates the
// numbers; this example itself only asserts the host numbers are sane.

use std::fmt;
use std::fs::{self, OpenOptions};
use std::hint::black_box;
use std::io::Write as _;
use std::time::Instant;

const USAGE: &str = "usage: overheadbench [--bench syscall|mem|io|all] [--iters N]";

// Defaults are chosen so each bench runs for roughly tens to a few hundred
// milliseconds on a modern host; identical across all three languages.
const SYSCALL_DEFAULT_ITERS: u64 = 200_000;
const MEM_DEFAULT_PASSES: u64 = 16;
const MEM_BUF_BYTES: u64 = 128 * 1024 * 1024; // 128 MiB
const IO_DEFAULT_ITERS: u64 = 200;
const IO_FILE_BYTES: usize = 4096;

#[derive(Clone, Copy, PartialEq)]
enum Bench {
    Syscall,
    Mem,
    Io,
    All,
}

struct Config {
    bench: Bench,
    iters: Option<u64>,
}

#[derive(Debug)]
struct ArgError(String);

impl fmt::Display for ArgError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

fn parse_args(args: &[String]) -> Result<Config, ArgError> {
    let mut cfg = Config {
        bench: Bench::All,
        iters: None,
    };
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--bench" => {
                i += 1;
                let v = args
                    .get(i)
                    .ok_or_else(|| ArgError("--bench requires a value".into()))?;
                cfg.bench = match v.as_str() {
                    "syscall" => Bench::Syscall,
                    "mem" => Bench::Mem,
                    "io" => Bench::Io,
                    "all" => Bench::All,
                    other => return Err(ArgError(format!("unknown --bench value: {other}"))),
                };
            }
            "--iters" => {
                i += 1;
                let v = args
                    .get(i)
                    .ok_or_else(|| ArgError("--iters requires a value".into()))?;
                let n: u64 = v.parse().map_err(|_| {
                    ArgError(format!("--iters value must be a positive integer: {v}"))
                })?;
                if n == 0 {
                    return Err(ArgError("--iters value must be >= 1".into()));
                }
                cfg.iters = Some(n);
            }
            other => return Err(ArgError(format!("unknown argument: {other}"))),
        }
        i += 1;
    }
    Ok(cfg)
}

/// Tight getpid(2) loop, ns/call.
fn bench_syscall(iters: u64) -> f64 {
    let start = Instant::now();
    for _ in 0..iters {
        // rustix::process::getpid issues the raw getpid(2) syscall every
        // call — every iteration really crosses into the kernel, the
        // syscall-boundary cost this chapter contrasts against VM (vmexit
        // trap) and container (near-zero extra) overhead.
        black_box(rustix::process::getpid());
    }
    let mut ns = start.elapsed().as_nanos() as f64;
    if ns <= 0.0 {
        ns = 1.0;
    }
    ns / iters as f64
}

/// Sequential read bandwidth over a big buffer, GB/s.
fn bench_mem(passes: u64) -> f64 {
    let words = (MEM_BUF_BYTES / 8) as usize;
    let mut buf = vec![0u64; words];
    for (i, w) in buf.iter_mut().enumerate() {
        *w = (i as u64).wrapping_mul(2654435761);
    }

    let mut sum: u64 = 0;
    let start = Instant::now();
    for _ in 0..passes {
        for &w in &buf {
            sum = sum.wrapping_add(w);
        }
    }
    let mut secs = start.elapsed().as_secs_f64();
    if secs <= 0.0 {
        secs = 1e-9;
    }

    // Consume `sum` after the clock stops so the compiler cannot prove the
    // whole scan is dead and elide it; this doesn't affect the measured time.
    black_box(sum);

    let bytes = MEM_BUF_BYTES as f64 * passes as f64;
    bytes / secs / 1e9
}

/// Create/write/fsync/unlink loop, ops/s.
fn bench_io(iters: u64) -> Result<f64, String> {
    // Relative to the CWD (the demo.sh working directory), not the system
    // temp dir: on many dev hosts that's tmpfs, where fsync is nearly free
    // and the number stops meaning anything as "disk IO overhead". The
    // example directory itself is normally on a real filesystem, which is
    // the point of comparing this number across host/VM/container.
    let dir = std::path::PathBuf::from(format!("overheadbench-io-{}", std::process::id()));
    fs::create_dir_all(&dir).map_err(|e| format!("mkdir: {e}"))?;
    let path = dir.join("probe");
    let payload = vec![b'x'; IO_FILE_BYTES];

    let cleanup = |p: &std::path::Path, d: &std::path::Path| {
        let _ = fs::remove_file(p);
        let _ = fs::remove_dir(d);
    };

    let start = Instant::now();
    for _ in 0..iters {
        let mut f = match OpenOptions::new()
            .create(true)
            .write(true)
            .truncate(true)
            .open(&path)
        {
            Ok(f) => f,
            Err(e) => {
                cleanup(&path, &dir);
                return Err(format!("open: {e}"));
            }
        };
        if let Err(e) = f.write_all(&payload) {
            cleanup(&path, &dir);
            return Err(format!("write: {e}"));
        }
        if let Err(e) = f.sync_all() {
            cleanup(&path, &dir);
            return Err(format!("fsync: {e}"));
        }
        drop(f);
        if let Err(e) = fs::remove_file(&path) {
            cleanup(&path, &dir);
            return Err(format!("unlink: {e}"));
        }
    }
    let mut secs = start.elapsed().as_secs_f64();
    if secs <= 0.0 {
        secs = 1e-9;
    }
    let _ = fs::remove_dir(&dir);
    Ok(iters as f64 / secs)
}

struct Row {
    name: &'static str,
    unit: &'static str,
    value: f64,
}

fn run_benches(cfg: &Config) -> Result<Vec<Row>, String> {
    let want_syscall = cfg.bench == Bench::Syscall || cfg.bench == Bench::All;
    let want_mem = cfg.bench == Bench::Mem || cfg.bench == Bench::All;
    let want_io = cfg.bench == Bench::Io || cfg.bench == Bench::All;

    let mut rows = Vec::new();
    if want_syscall {
        let iters = cfg.iters.unwrap_or(SYSCALL_DEFAULT_ITERS);
        rows.push(Row {
            name: "syscall",
            unit: "ns/call",
            value: bench_syscall(iters),
        });
    }
    if want_mem {
        let passes = cfg.iters.unwrap_or(MEM_DEFAULT_PASSES);
        rows.push(Row {
            name: "mem",
            unit: "GB/s",
            value: bench_mem(passes),
        });
    }
    if want_io {
        let iters = cfg.iters.unwrap_or(IO_DEFAULT_ITERS);
        let value = bench_io(iters).map_err(|e| format!("io bench failed: {e}"))?;
        rows.push(Row {
            name: "io",
            unit: "ops/s",
            value,
        });
    }
    Ok(rows)
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();

    let cfg = match parse_args(&args) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{e}");
            eprintln!("{USAGE}");
            std::process::exit(2);
        }
    };

    let rows = match run_benches(&cfg) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("{e}");
            std::process::exit(1);
        }
    };

    for row in &rows {
        println!(
            "bench={} metric={:.2} unit={}",
            row.name, row.value, row.unit
        );
    }
}
